### DISCUSSION TITLE
Telescope plugin weird config behavior

### DISCUSSION THREAD

**Discussion by kapral18** [LazyVim/LazyVim#1811] [Q&A] [answered]
I tried to refactor my telescope config on top of the lazyvim defaults but most customized behavior seems to not work at all. 

Please help me understand why. 

Here is my config: https://github.com/kapral18/config.nvim/blob/56b3aaa6fdd3fc282df9711ec1a14b7398058d3f/lua/plugins/telescope.lua

The following behaviors do not work:

1. Disabled keymaps in `keys` are not actually being disabled and continue to respond
2. `"<leader><space>"` mapping doesn't work at all
3. `<a-h>` and `<a-i>` remaps don't work
4. `live_grep_args` mappings don't work as well

It might look like the entire plugin customization doesn't work, but some things do work, like for example the mapping for `live_grep_args` does actually get overwritten and shows as expected.

Thanks

**Comment by dpetka2001** [comment, marked as answer]
1. Disabled mappings
```lua
{ "<leader>sg", false },
    { "<leader>sG", false },
    { "<leader>ff", false },
    { "<leader>fF", false },
```
work as expected for me.
2. `<leader><space>` doesn't work for me either.
3. remap `{ "<leader>gb", ":Telescope git_branches<CR>", desc = "Branches" },` works fine for me
4. live_grep mappings: for me the following [gist](https://gist.github.com/dpetka2001/a289752c3fcaf308b88127609beb3698) worked. Something seems to be happening with `Util.on_load` and it mustn't be keeping the correct order. I'm not sure if this function gets called after the setup() of the plugin finishes or not. It just mentions `added util.on_load to execute code when a plugin loads`. I have searched the repo for where it's being used and it was used in `aerial` and `fzf`, both of which don't set up any options in the telescope `opts.extensions`. But since `live_grep` sets up options in telescope's `opts.extensions` its setup function must for sure be called after the setup function of telescope. Maybe Folke would be able to give you a more concrete answer, if you'd like to ping him. I would like to refrain for pinging him myself, cuz I've already bothered him enough times the past few days. Sorry for leaving this up to you.

**Reply by kapral18** [reply]
My bad. By "remaps" I meant the `<a-i>` and `<a-h>` remaps I did within `opts` not the git branch. Apparently forgot to wrap in backticks and md rendered them as empty string.

Btw, have you tried my setup locally or did you reconstruct this within yours?

**Reply by dpetka2001** [reply]
I copy/pasted your `telescope.lua` file in a new Lazyvim starter template to work only with as default setup as possible. Your `<a-i>`/`<a-h>` remaps have something to do with the local function you created. I used the default [find_files_with_hidden](https://github.com/LazyVim/LazyVim/blob/0bb628053beec25efbc5bd1e8600ef039fa75a20/lua/lazyvim/plugins/editor.lua#L199-L203) that Lazyvim uses and it works correctly. So the keymap registers, it's just that your function must have an error.

**Reply by dpetka2001** [reply]
I should clarify that I changed your default settings for `grep_with_args` to not use `Util.on_load()` and instead configured it in the `config` function of Telescope like shown in the gist I linked to in my previous post, because it wouldn't work otherwise.

**Reply by kapral18** [reply]
I mean, I just reused the approach from here https://github.com/LazyVim/LazyVim/blob/0bb628053beec25efbc5bd1e8600ef039fa75a20/lua/lazyvim/plugins/editor.lua#L123C1-L125

Thanks for the gist btw, will check

**Reply by kapral18** [reply]
I check your config indeed with this setup remaps work, but my `<leader><space>` mapping that I enabled to double check doesn't. Still yields nothing.

**Reply by kapral18** [reply]
@folke could you please assist? Why is Util.on_load working in lazyvim code but messing up stuff in my abstraction layer? And do you know why my `<leader><space>` mapping function even with @dpetka2001's [correction](https://gist.github.com/dpetka2001/a289752c3fcaf308b88127609beb3698) not working?

**Reply by folke** [reply]
Leader space is a global mapping, so you need to add it to keymaps.lua.

On_load is an internal function, so don't use it if you don't know what you're doing

**Reply by kapral18** [reply]
@folke 

What is global? If I check your docs it says mappings that are "always active"

But at the same time I check https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-adding--disabling-plugin-keymaps

I see the mappings from the same exact list where `<leader><space>` is defined.

https://github.com/LazyVim/LazyVim/blob/fa3170d422f3c661d0411472c96f92e5324dc281/lua/lazyvim/plugins/editor.lua#L133

Could you clarify how can I definitively differentiate between global and  non-global mappings in this context?

And let's say I move all these mappings into `keymaps.lua`, will plugins be loaded  by the time `keymaps.lua` is loaded so I can safely require them or should I create some alternative to Util.on_load?

**Reply by folke** [reply]
No, only that space mapping. The ones under keymaps that are not part of a plugin

**Reply by kapral18** [reply]
@folke do you mean `keymaps.lua` ? I don't see `<leader><space>` anywhere in [`keymaps.lua`](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/keymaps.lua)

Could you kindly show me where to find that non-plugin global mapping of `<leader><space>` besides the plugin specific one in editor.lua under telescope.nvim ?

**Reply by dpetka2001** [reply]
@kapral18 In line [41](https://github.com/kapral18/config.nvim/blob/56b3aaa6fdd3fc282df9711ec1a14b7398058d3f/lua/plugins/telescope.lua#L41C95-L41C95) add `()` at the end to call the function `Util.telescope`. That seemed to have fixed the issue with `<leader><space>` for me. So, it should be like
```lua
Util.telescope("find_files", { no_ignore = true, hidden = true, default_text = line })()
```

**Reply by kapral18** [reply]
@dpetka2001 wow thanks much for finding the culprit for that one.

**Reply by kapral18** [reply]
@dpetka2001 could you maybe  explain the difference between `<leader><space>` and other mappings alongside this [block](https://github.com/LazyVim/LazyVim/blob/fa3170d422f3c661d0411472c96f92e5324dc281/lua/lazyvim/plugins/editor.lua#L130-L183) within editor.lua? Why only `<leader><space>` is considered global out of this list?

**Reply by folke** [reply]
I was clearly wrong. It's a plugin mapping, not global. Your issue was indeed that you didn't call the returned telescope function

**Reply by dpetka2001** [reply]
Not sure. I haven't considered it global, since i saw it configured as a LazyKey handler inside the telescope spec. Not sure what @folke means. If he would like to elaborate on that, i too would be greatful, so that i can properly understand.

Edit: sorry @folke i was writing on my phone when you already answered.

**Reply by kapral18** [reply]
Thank you guys @folke and @dpetka2001 for clarifyng this. Appreciate it.

**Comment by jyuan0** [comment]
1. Should definitely work.
2. Try `Util.telescope("files", ...` instead of `"find_files"`.
3. Should definitely work.
4. Sorry, I don't use live_grep, so can't really can't help here.

If you're wondering if the `on_load` is affecting anything, you could just remove it and call `load_extension` in a plugin `config` function to test (call `setup` for Telescope as well if you do add a `config` function).

**Reply by kapral18** [reply]
@jyuan0 telescope("files") is preventing usage of builtin `find_files` and uses `git_files` instead for .git folders which is 99% of use cases for me. `git_files` doesn't seem to get covered by handy telescope mappings https://github.com/LazyVim/LazyVim/blob/0bb628053beec25efbc5bd1e8600ef039fa75a20/lua/lazyvim/plugins/editor.lua#L194-L203
which btw don't properly work for me either, because for example gitignored files don't show unless they are no-ignore-d and hidden at the same time so I tried to make that behavior by default for find_files and and onnly toggle parts off/on with those `<a-i>` and `<a-h>` mappings in my op

**Reply by dpetka2001** [reply]
I just left `<leader>ff` with default `Util.telescope("files")` and changed `<leader><space>` to `Util.telescope("find_files")` to be able to differentiate when I happen to be in a `.git` directory. Nothing wrong with that.

**Reply by jyuan0** [reply]
> @jyuan0 telescope("files") is preventing usage of builtin find_files and uses git_files instead for .git folders which is 99% of use cases for me. git_files doesn't seem to get covered by handy telescope mappings

Oh ok.

> I just left <leader>ff with default Util.telescope("files") and changed <leader><space> to Util.telescope("find_files") to be able to differentiate when I happen to be in a .git directory. Nothing wrong with that.

Think I'd just do this too.
